lib/run_part.c

```diff
@@ -51,47 +52,73 @@ int run_parts(const char *directory, const char *name, const char *action)
 	int n;
 	int execute_result = 0;
 
+	int dfd = open(directory, O_PATH | O_DIRECTORY | O_CLOEXEC);
+	if (dfd == -1) {
+		fprintf(shadow_logfd, "failed open directory %s: %s\n", directory, strerror(errno));
+		return (1);
+	}
+
+#ifdef HAVE_SCANDIRAT
+	scanlist = scandirat(dfd, ".", &namelist, 0, alphasort);
+#else
 	scanlist = scandir(directory, &namelist, 0, alphasort);
+#endif
 	if (scanlist<=0) {
+		(void) close(dfd);
 		return (0);
 	}
 
 	for (n=0; n<scanlist; n++) {
-		char         *s;
-		struct stat  sb;
-
-		if (asprintf(&s, "%s/%s", directory, namelist[n]->d_name) == -1) {
-			fprintf(shadow_logfd, "asprintf: %s\n", strerror(errno));
-			for (; n<scanlist; n++) {
-				free(namelist[n]);
+		struct stat sb;
```

Comment:
Please keep the 2 spaces before the variable name.  That will keep the diff smaller, and also have it easier to distinguish where the type name ends and the variable name starts.